### PR TITLE
Enable @vectorize for target dppy

### PR DESCRIPTION
**Pull request by reazulhoque**
This PR will allow users to decorate functions with `@vectorize`  which can be directly offloaded to target dppy.

This PR is a part of #277.

### CODE REVIEW

File: numba_dppy/compiler.py [hunk 1]
```diff
@@ -360,6 +358,87 @@ def __getitem__(self, args):
...
+        arr: Array_like object.
+
+    Retruns:
+        bool: True if array is device accessible, False otherwise.
+    """
+    rc = False
+    if hasattr(arr, "__sycl_usm_array_interface__"):
+        rc = True
+    elif hasattr(arr, "base"):
+        rc = hasattr(arr.base, "__sycl_usm_array_interface__")
```
Comment (PokhodenkoSA) [inline numba_dppy/compiler.py:375]: `"__sycl_usm_array_interface__"` constant duplication.

File: numba_dppy/compiler.py [hunk 2]
```diff
@@ -360,6 +358,87 @@ def __getitem__(self, args):
...
+    """
+    Function to determine if a given array is SYCL device accessible.
+
+    Args:
+        arr: Array_like object.
+
+    Retruns:
+        bool: True if array is device accessible, False otherwise.
+    """
+    rc = False
```
Comment (PokhodenkoSA) [inline numba_dppy/compiler.py:371]: What is `rc`? May be `result`?

File: numba_dppy/compiler.py [hunk 3]
```diff
@@ -360,6 +358,87 @@ def __getitem__(self, args):
...
+    if hasattr(arr, "__sycl_usm_array_interface__"):
+        rc = True
```
Comment (PokhodenkoSA) [inline numba_dppy/compiler.py:373]: Why not just return `True` from here?

File: numba_dppy/compiler.py [hunk 4]
```diff
@@ -360,6 +358,87 @@ def __getitem__(self, args):
...
+        raise TypeError("dtype has to be of type np.dtype. Got %s" % (type(dtype)))
+    if not isinstance(queue, dpctl._sycl_queue.SyclQueue):
+        raise TypeError(
+            "queue has to be of dpctl._sycl_queue.SyclQueue type. Got %s"
+            % (type(queue))
+        )
+
+    size = 1
+    for i in shape:
+        size *= i
```
Comment (PokhodenkoSA) [inline numba_dppy/compiler.py:410]:
```suggestion
    size = np.prod(shape)
```

File: numba_dppy/compiler.py [hunk 5]
```diff
@@ -495,15 +574,12 @@ def _unpack_argument(
...
+            if is_device_array(val):
                 self._unpack_device_array_argument(val, kernelargs)
             else:
                 default_behavior = self.check_for_invalid_access_type(access_type)
 
-                usm_buf = dpctl_mem.MemoryUSMShared(
-                    val.size * val.dtype.itemsize, queue=sycl_queue
-                )
-                usm_ndarr = np.ndarray(val.shape, buffer=usm_buf, dtype=val.dtype)
+                usm_ndarr = device_array(val.shape, val.dtype, sycl_queue)
```
Comment (PokhodenkoSA) [inline numba_dppy/compiler.py:582]: In Numpy world this is similar to `empty_like(val, sycl_queue)`.

File: numba_dppy/compiler.py [hunk 6]
```diff
@@ -512,7 +588,7 @@ def _unpack_argument(
                 ):
                     np.copyto(usm_ndarr, val)
 
-                device_arrs[-1] = (usm_buf, usm_ndarr, val)
+                device_arrs[-1] = (None, usm_ndarr, val)
```
Comment (PokhodenkoSA) [inline numba_dppy/compiler.py:591]: Why `None` now?

**Review by PokhodenkoSA — COMMENTED** [review]
See my remarks.

File: numba_dppy/compiler.py (hunk 2 above)
Comment (reazulhoque) [reply numba_dppy/compiler.py:373]: return code

File: numba_dppy/compiler.py (hunk 3 above)
Comment (reazulhoque) [reply numba_dppy/compiler.py:375]: What do we gain?

File: numba_dppy/compiler.py [hunk 7]
```diff
@@ -362,6 +360,85 @@ def __getitem__(self, args):
...
+    rc = False
+    if hasattr(arr, "__sycl_usm_array_interface__"):
+        rc = True
+    elif hasattr(arr, "base"):
+        rc = hasattr(arr.base, "__sycl_usm_array_interface__")
+
+    return rc
+
+
+def device_array(shape, dtype, queue):
```
Comment (diptorupd) [inline numba_dppy/compiler.py:382]: The function should be renamed. How about `as_device_array`?

**Comment by diptorupd** [comment]
@reazulhoque A general question about the changes. Should we create a `DeviceArray` class and convert all the functions into methods and properties? Also, why are all these functions in `compiler.py`? It might be worthwhile to have them in a separate module to not make our `compiler.py` bloated.

File: numba_dppy/compiler.py [hunk 8]
```diff
@@ -34,6 +33,10 @@
 from numba_dppy import config
 from numba_dppy.driver import USMNdArrayType
 from numba_dppy.dppy_array_type import DPPYArray
+from numba_dppy.utils.array_utils import (
+    _is_device_accessible_array,
```
Comment (diptorupd) [inline numba_dppy/compiler.py:37]: What is the correct Pythonic style? If these functions are public functions in `array_utils` they should not have a leading `_`.

File: numba_dppy/compiler.py [hunk 9]
```diff
@@ -455,7 +456,7 @@ def _pack_argument(self, ty, val, sycl_queue, device_arr, access_type):
             # we get the date back to host if have created a
```
Comment (diptorupd) [inline numba_dppy/compiler.py:456]: s/date/data

Comment (diptorupd) [reply numba_dppy/compiler.py:456]: The comment should overall be updated. I think it is carry over from when we did have a `device_array`. Now that there is no such type, the comment should be cleaned up.

Stylistic nitpick: Comments can be 72 char long per PEP-8.

File: numba_dppy/compiler.py [hunk 10]
```diff
@@ -455,7 +456,7 @@ def _pack_argument(self, ty, val, sycl_queue, device_arr, access_type):
...
             # device_array or if access_type of this device_array
             # is not of type read_only and read_write
-            usm_buf, usm_ndarr, orig_ndarray = device_arr
+            usm_ndarr, orig_ndarray = device_arr
             np.copyto(orig_ndarray, usm_ndarr)
```
Comment (diptorupd) [inline numba_dppy/compiler.py:460]: To be consistent with the `orig_ndarray`, let us do s/usm_ndarr/usm_array. I do not want to call it `usm_ndarray` as that can be confusing given `usm_ndarray` is an actual data type in dpctl.

File: numba_dppy/compiler.py [hunk 11]
```diff
@@ -614,8 +614,7 @@ def specialize(self, argtypes, queue):
         kernel = None
         # we were previously using the _env_ptr of the device_env, the sycl_queue
         # should be sufficient to cache the compiled kernel for now, but we should
-        # use the device type to cache such kernels
-        # key_definitions = (self.sycl_queue, argtypes)
```
Comment (diptorupd) [inline numba_dppy/compiler.py:618]: 👍

File: numba_dppy/tests/test_vectorize.py [hunk 12]
```diff
@@ -14,41 +14,101 @@
 # limitations under the License.
 
 import numpy as np
-from numba import njit, vectorize
+from numba import njit, vectorize, int32, float32, int64, float64
 from . import _helper
 import dpctl
-import unittest
+import pytest
```
Comment (diptorupd) [inline numba_dppy/tests/test_vectorize.py:20]: 👍

File: numba_dppy/utils/array_utils.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import dpctl
+import dpctl.memory as dpctl_mem
+
+DEVICE_ARRAY_ATTR = "__sycl_usm_array_interface__"
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:19]: Let us call it `SYCL_USM_ARRAY_INTERFACE_ATTR`.

File: numba_dppy/utils/array_utils.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def _is_device_accessible_array(arr):
+    """
+    Function to determine if a given array is SYCL device accessible.
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:24]: Under pydocstyle guide the comment should be "Determine if a given array is SYCL device accessible."

File: numba_dppy/utils/array_utils.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+    Args:
+        arr: array_like object.
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:27]: what is `array_like`? Is it a numpy type? If not, remove the underscore.

File: numba_dppy/utils/array_utils.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+    Retruns:
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:29]: typo

File: numba_dppy/utils/array_utils.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+        bool: True if array is device accessible, False otherwise.
+    """
+    if hasattr(arr, DEVICE_ARRAY_ATTR):
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:32]: IMO just checking if the attribute exists is not enough. Look at this Pytorch test case https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/test/test_numba_integration.py#L84.

We should check the contents of the dictionary and the version of the interface.

File: numba_dppy/utils/array_utils.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    elif hasattr(arr, "base"):
+        if hasattr(arr.base, DEVICE_ARRAY_ATTR):
+            return True
+
+    return False
+
+
+def _as_device_accessible_array(shape, dtype, queue):
+    """
+    Allocate USM shared buffer and create a np.ndarray
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:43]: Per pydocstyle should be in one line.

File: numba_dppy/utils/array_utils.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        np.ndarray: NumPy array created using a USM shared buffer.
+
+    Raises:
+        TypeError: if any argument is not of permitted type.
+    """
+    if not (isinstance(shape, int) or isinstance(shape, tuple)):
+        raise TypeError(
+            "Shape has to be a integer or tuple of integers. Got %s" % (type(shape))
+        )
+    if not isinstance(dtype, np.dtype):
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:61]: What if we pass in some NumPy type that is not supported by us, e.g. np,half?

File: numba_dppy/utils/array_utils.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        raise TypeError("dtype has to be of type np.dtype. Got %s" % (type(dtype)))
+    if not isinstance(queue, dpctl._sycl_queue.SyclQueue):
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:63]: you should be able to access `SyclQueue` directly from `dpctl` namespace.

File: numba_dppy/utils/array_utils.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    if hasattr(arr, DEVICE_ARRAY_ATTR):
+        return True
+    elif hasattr(arr, "base"):
+        if hasattr(arr.base, DEVICE_ARRAY_ATTR):
+            return True
+
+    return False
+
+
+def _as_device_accessible_array(shape, dtype, queue):
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:41]: let us improve the name to `as_usm_backed_array`?

File: numba_dppy/utils/array_utils.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+
+    size = np.prod(shape)
+
+    usm_buf = dpctl_mem.MemoryUSMShared(size * dtype.itemsize, queue=queue)
+    usm_ndarr = np.ndarray(shape, buffer=usm_buf, dtype=dtype)
+
+    return usm_ndarr
+
+
+def _to_device_accessible_array(hostary, queue):
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:77]: again the `device` word is not descriptive enough.

Comment (diptorupd) [reply numba_dppy/utils/array_utils.py:77]: `hostary` reads weird.

File: numba_dppy/utils/array_utils.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+    """
+    Create a np.ndarray using USM shared buffer and copy the data from
+    provided array to the newly allocated buffer.
+
+    Args:
+        hostary (np.ndarray): Array data will be copied from.
+        queue (dpctl._sycl_queue.SyclQueue): SYCL queue used to create the buffer.
+
+    Returns:
+        np.ndarray: Device accessible NumPy array with same data as hostary.
```
Comment (diptorupd) [inline numba_dppy/utils/array_utils.py:87]: s/np.array/numpy.array/ Someone reading a generated doc will have no understanding what the alias stands for in the documentation.

**Comment by diptorupd** [comment]
I suggest you break off the array_utils into a separate PR to keep things small.

File: numba_dppy/utils/array_utils.py (hunk 21 above)
Comment (reazulhoque) [reply numba_dppy/utils/array_utils.py:41]: May be `usm_backed_ndarray`.

**Comment by reazulhoque** [comment]
Closing as https://github.com/IntelPython/numba-dppy/pull/497 supercedes this PR.

**Pull request closed without merging**